### PR TITLE
Interface extends interfaces

### LINKED ISSUE #24: Interface should be able to enforce other interfaces

### PR DESCRIPTION
**Pull request by ajeet-smpx**
- fixes #24 
- Extend interfaces by defining `extends` property in the interface.
  Ex.  
	```js
	const InterfaceB = {
		graphql: 'interface',
		...
	}

	const InterfaceA = {
		graphql: 'interface',
		extends: ['InterfaceB'],
		...
	}
	```

### CODE REVIEW

File: src/Schema.js [hunk 1]
```diff
@@ -59,6 +59,33 @@ function getMergedTypeFieldsWithInterfaces(schema, type) {
...
 
+function collectSingleInterfaceDependencies(schema, interfaceName) {
+	const dependencies = [];
+	const processQueue = [interfaceName];
+
+	while (processQueue.length) {
+		const top = processQueue.shift();
+		const _interface = schema.interfaces[top];
+
+		if (_interface === undefined) throw new Error(`Interface "${top}" is not defined`);
```
Comment (rohit-gohri) [inline src/Schema.js]: I just realised why we shouldn't throw if we don't find an interface. If we are declaring multiple schemas: `admin`, `public` then an interface might only exist in one of them but the type will extend it still

Comment (rohit-gohri) [reply src/Schema.js]: Maybe we should also check that condition? If `interface.schema.includes(schema.name)`?

File: src/Schema.js [hunk 2]
```diff
@@ -433,6 +432,39 @@ class Schema {
 		});
 	}
 
+	collectSingleInterfaceDependencies(schema, interfaceName) {
+		let interfaces = this.schemas[this.defaultSchemaName].interfaces;
+		interfaces = interfaces.map(i => i.name);
```
Comment (rohit-gohri) [inline src/Schema.js]: Wouldn't `interfaces` be an object type?

Comment (rohit-gohri) [inline src/Schema.js]: Could rename this to `allInterfaces`

File: src/Schema.js [hunk 3]
```diff
@@ -705,10 +737,18 @@ class Schema {
 		const resolveType = this.resolvers[schemaItem.name] &&
 			this.resolvers[schemaItem.name].__resolveType;
 
+		const dependencies = this.collectSingleInterfaceDependencies(schema, schemaItem.name);
+
+		const fields = {};
+		for (let i = dependencies.length - 1; i >= 0; i--) {
```
Comment (rohit-gohri) [inline src/Schema.js:757]: Add comment here why reverse iteration is being done

File: src/Schema.js (hunk 2 above)
Comment (ajeet-smpx) [reply src/Schema.js]: Oh.. :P,
Fixing it

File: src/Schema.js [hunk 4]
```diff
@@ -59,6 +59,31 @@ function getMergedTypeFieldsWithInterfaces(schema, type) {
...
+	// Already Processing
+	if (state[interfaceName] === 1) throw new Error(`Cyclic dependencies at interface "${interfaceName}"`);
+	// Already Processed
+	else if (state[interfaceName] === 2) return [];
+
+	const dependencies = [];
+	const _interface = schema.interfaces[interfaceName];
+
+	if (_interface === undefined) {
+		if (allInterfaces.includes(interfaceName)) return [];
```
Comment (rohit-gohri) [inline src/Schema.js:73]: Comment can be added here?

**Review by rohit-gohri — APPROVED** [review]

**Pull request merged**